### PR TITLE
GSPO implementation

### LINKED ISSUE #297: [FSDP] Implement GSPO algorithm
The FSDP backend currently lacks support for the GSPO ([Generalized Second-Order Policy Optimization](https://arxiv.org/abs/2507.18071)) algorithm. This is necessary for MoE training.

### PR DESCRIPTION
**Pull request by ppraneth**
# GSPO Implementation 

**Closes #297**

## Description

This pull request introduces a complete implementation of the **Group Sequence Policy Optimization (GSPO)** algorithm, as detailed in the paper "Group Sequence Policy Optimization" ([[arXiv:2507.18071v2](https://arxiv.org/abs/2507.18071v2)](https://arxiv.org/abs/2507.18071v2)) This addition is critical for improving the stability and performance of reinforcement learning, especially for large-scale Mixture-of-Experts (MoE) models where existing algorithms like GRPO can exhibit instability.

This implementation includes both the standard GSPO algorithm and its flexible GSPO-token variant, providing users with robust new tools for RL training.

## Key Changes and Implementation Details

The implementation is broken down into three main components, directly following the methodology outlined in the paper.

### 1. Group-Based Advantage Normalization

As specified in the paper, both GRPO and GSPO utilize a group-based advantage estimation to provide a more stable reward signal. The advantage $\hat{A}_i$ for a sequence $y_i$ is calculated by normalizing its reward against the mean and standard deviation of rewards from other sequences ($G$) generated for the same prompt.

**Formula (Equation 6):**

$$\hat{A}_i = \frac{r(x, y_i) - \text{mean}(\{r(x, y_j)\}_{j=1}^{G})}{\text{std}(\{r(x, y_j)\}_{j=1}^{G})}$$

**Implementation:** In `slime/utils/ppo_utils.py`:

```python
def normalize_advantages_in_groups(rewards: torch.Tensor, n_samples_per_prompt: int) -> torch.Tensor:
    # ... implementation ...
```

### 2. Standard GSPO: Sequence-Level Optimization

The core innovation of GSPO is its use of a sequence-level importance ratio, which aligns the optimization objective with the sequence-level reward signal. This avoids the high variance associated with GRPO's token-level ratios.

**Importance Ratio Formula (Equation 7):** 

The ratio $s_i(\theta)$ is the length-normalized sequence likelihood ratio, equivalent to the geometric mean of the token-level ratios.

$$s_i(\theta) = \left( \frac{\pi_\theta(y_i|x)}{\pi_{\theta_{old}}(y_i|x)} \right)^{\frac{1}{|y_i|}}$$

**Implementation:** Implemented in the new function `get_sequence_level_ratio` in `slime/utils/ppo_utils.py`.

### 3. GSPO-token Variant for Finer-Grained Control

The GSPO-token variant is implemented for advanced use cases (e.g., multi-turn RL) that require token-level advantage customization. This variant uses a modified importance ratio combining a detached sequence-level ratio with the standard token-level ratio.

**Importance Ratio Formula (Equation 14):**

$s_{i,t}(\theta) = \text{sg}[s_i(\theta)] \cdot \frac{\pi_\theta(y_{i,t}|x, y_{i,<t})}{\text{sg}[\pi_\theta(y_{i,t}|x, y_{i,<t})]}$

where $\text{sg}[\cdot]$ denotes the stop-gradient operation.

**Implementation:** Implemented in the new function `get_gspo_token_ratio` in `slime/utils/ppo_utils.py`.

### 4. Integration into FSDP Backend

The new functionalities are integrated into the `FSDPTrainRayActor` in `slime/backends/fsdp_utils/actor.py`:

- Added command-line arguments `--advantage-estimator gspo` and `--advantage-estimator gspo-token`
- A conditional block selects the appropriate ratio calculation (`get_sequence_level_ratio` or `get_gspo_token_ratio`) based on the chosen algorithm
- The output ratio is transformed to `-torch.log(ratio)` before being passed to the `compute_policy_loss` function, ensuring alignment with the existing PPO objective

## How to Use

To use the new algorithms, set the `--advantage-estimator` flag during training:

**For standard GSPO:**
```bash
--advantage-estimator gspo
```

**For the GSPO-token variant:**
```bash
--advantage-estimator gspo-token
```

### CODE REVIEW

**Comment by ppraneth** [comment]
Also added gspo-token 
@zhuzilin can you take a look at this pr

**Comment by ppraneth** [comment]
@zhuzilin Can you take a look at this pr this adds **gspo** and **gspo-tokens**

**Comment by zhuzilin** [comment]
this is too much code for this feature... I would expect an pr with around 20 loc....
